File: app/src/main/java/com/starsep/sokoban/gamelogic/Move.java
package com.starsep.sokoban.gamelogic;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;

import com.starsep.sokoban.res.Textures;

public class Move {
	public static class UnknownMoveException extends Exception { }

	@NonNull public static final Move DOWN = new Move(Direction.DOWN, false);
	@NonNull public static final Move UP = new Move(Direction.UP, false);
	@NonNull public static final Move LEFT = new Move(Direction.LEFT, false);
	@NonNull public static final Move RIGHT = new Move(Direction.RIGHT, false);
	@NonNull public static final Move PUSH_DOWN = new Move(Direction.DOWN, true);
	@NonNull public static final Move PUSH_UP = new Move(Direction.UP, true);
	@NonNull public static final Move PUSH_LEFT = new Move(Direction.LEFT, true);
	@NonNull public static final Move PUSH_RIGHT = new Move(Direction.RIGHT, true);

	@NonNull private final Direction direction;
	private final boolean pushed;

	@NonNull
	public static Move make_push(@NonNull Move move) {
		switch (move.direction) {
			case LEFT:
				return PUSH_LEFT;
			case RIGHT:
				return PUSH_RIGHT;
			case UP:
				return PUSH_UP;
			case DOWN:
				return PUSH_DOWN;
			default:
				return PUSH_DOWN;
		}
	}

	@NonNull
	public Move reverse() {
		switch (direction) {
			case LEFT:
				return push() ? PUSH_RIGHT : RIGHT;
			case RIGHT:
				return push() ? PUSH_LEFT : LEFT;
			case DOWN:
				return push() ? PUSH_UP : UP;
			default:
			case UP:
				return push() ? PUSH_DOWN : DOWN;
		}
	}

	@NonNull
	public Bitmap heroTexture() {
		switch (direction) {
			case LEFT:
				return Textures.heroLeft();
			case RIGHT:
				return Textures.heroRight();
			case UP:
				return Textures.heroUp();
			default:
			case DOWN:
				return Textures.heroDown();
		}
	}

	public enum Direction {
		LEFT,
		RIGHT,
		DOWN,
		UP
	}

	private Move(@NonNull Direction direction, boolean pushed) {
		this.direction = direction;
		this.pushed = pushed;
	}

	public int dx() {
		switch (direction) {
			case LEFT:
				return -1;
			case RIGHT:
				return 1;
			case UP:
			case DOWN:
			default:
				return 0;
		}
	}

	public int dy() {
		switch (direction) {
			case LEFT:
			case RIGHT:
				return 0;
			case UP:
				return -1;
			case DOWN:
			default:
				return 1;
		}
	}

	public boolean push() {
		return pushed;
	}

	@Override
	@NonNull
	public String toString() {
		return "" + toChar();
	}


	public char toChar() {
		switch (direction) {
			case LEFT:
				return push() ? 'L' : 'l';
			case RIGHT:
				return push() ? 'R' : 'r';
			case DOWN:
				return push() ? 'D' : 'd';
			default:
			case UP:
				return push() ? 'U' : 'u';
		}
	}

	@NonNull
	public static Move fromChar(char c) throws UnknownMoveException {
		switch (c) {
			case 'l':
				return LEFT;
			case 'L':
				return PUSH_LEFT;
			case 'r':
				return RIGHT;
			case 'R':
				return PUSH_RIGHT;
			case 'u':
				return UP;
			case 'U':
				return PUSH_UP;
			case 'd':
				return DOWN;
			case 'D':
				return PUSH_DOWN;
			default:
				throw new UnknownMoveException();
		}
	}
}

File: app/src/main/java/com/starsep/sokoban/release/gamelogic/Move.kt
package com.starsep.sokoban.release.gamelogic

import com.starsep.sokoban.release.res.Textures

class Move private constructor(private val direction: Direction, val push: Boolean) {
    fun reverse() = when (direction) {
        Direction.LEFT -> if (push) PUSH_RIGHT else RIGHT
        Direction.RIGHT -> if (push) PUSH_LEFT else LEFT
        Direction.DOWN -> if (push) PUSH_UP else UP
        Direction.UP -> if (push) PUSH_DOWN else DOWN
    }

    fun heroTexture() = when (direction) {
        Direction.LEFT -> Textures.heroLeft
        Direction.RIGHT -> Textures.heroRight
        Direction.UP -> Textures.heroUp
        Direction.DOWN -> Textures.heroDown
    }

    fun dx() = when (direction) {
        Direction.LEFT -> -1
        Direction.RIGHT -> 1
        Direction.UP, Direction.DOWN -> 0
    }

    fun dy() = when (direction) {
        Direction.LEFT, Direction.RIGHT -> 0
        Direction.UP -> -1
        Direction.DOWN -> 1
    }

    override fun toString() = "" + toChar()

    fun toChar() = when (direction) {
        Direction.LEFT -> if (push) 'L' else 'l'
        Direction.RIGHT -> if (push) 'R' else 'r'
        Direction.DOWN -> if (push) 'D' else 'd'
        Direction.UP -> if (push) 'U' else 'u'
    }

    enum class Direction {
        LEFT,
        RIGHT,
        DOWN,
        UP
    }

    class UnknownMoveException : Exception()

    companion object {
        val DOWN = Move(Direction.DOWN, false)
        val UP = Move(Direction.UP, false)
        val LEFT = Move(Direction.LEFT, false)
        val RIGHT = Move(Direction.RIGHT, false)
        val PUSH_DOWN = Move(Direction.DOWN, true)
        val PUSH_UP = Move(Direction.UP, true)
        val PUSH_LEFT = Move(Direction.LEFT, true)
        val PUSH_RIGHT = Move(Direction.RIGHT, true)

        fun makePush(move: Move) = when (move.direction) {
            Direction.LEFT -> PUSH_LEFT
            Direction.RIGHT -> PUSH_RIGHT
            Direction.UP -> PUSH_UP
            Direction.DOWN -> PUSH_DOWN
        }

        @Throws(UnknownMoveException::class)
        fun fromChar(c: Char) = when (c) {
            'l' -> LEFT
            'L' -> PUSH_LEFT
            'r' -> RIGHT
            'R' -> PUSH_RIGHT
            'u' -> UP
            'U' -> PUSH_UP
            'd' -> DOWN
            'D' -> PUSH_DOWN
            else -> throw UnknownMoveException().also {
                println("ASDASDSADASD $c")
            }
        }
    }

    fun toPosition() = Position(dy(), dx())
}
